RecoLocalFastTime/FTLClusterizer/src/MTDThresholdClusterizer.cc

```diff
@@ -0,0 +1,276 @@
+// Our own includes
+#include "RecoLocalFastTime/FTLClusterizer/interface/MTDArrayBuffer.h"
+#include "RecoLocalFastTime/FTLClusterizer/interface/MTDThresholdClusterizer.h"
+
+#include "DataFormats/ForwardDetId/interface/BTLDetId.h"
+#include "DataFormats/ForwardDetId/interface/ETLDetId.h"
+
+#include "Geometry/MTDGeometryBuilder/interface/ProxyMTDTopology.h"
+#include "Geometry/MTDGeometryBuilder/interface/RectangularMTDTopology.h"
+
+// STL
+#include <stack>
+#include <vector>
+#include <iostream>
+#include <atomic>
+using namespace std;
+
+//#define DEBUG_ENABLED 
+#ifdef DEBUG_ENABLED
+#define DEBUG(x) do { std::cout << x << std::endl; } while (0)
+#else
+#define DEBUG(x)
+#endif
+
+//----------------------------------------------------------------------------
+//! Constructor: 
+//----------------------------------------------------------------------------
+MTDThresholdClusterizer::MTDThresholdClusterizer
+  (edm::ParameterSet const& conf) :
+    // Get energy thresholds 
+    theHitThreshold( conf.getParameter<double>("HitThreshold") ),
+    theSeedThreshold( conf.getParameter<double>("SeedThreshold") ),
+    theClusterThreshold( conf.getParameter<double>("ClusterThreshold") ),
+    theNumOfRows(0), theNumOfCols(0), theCurrentId(0), 
+    theBuffer(theNumOfRows, theNumOfCols ),
+    bufferAlreadySet(false)
+{
+}
+/////////////////////////////////////////////////////////////////////////////
+MTDThresholdClusterizer::~MTDThresholdClusterizer() {}
+
+
+// Configuration descriptions
+void
+MTDThresholdClusterizer::fillDescriptions(edm::ParameterSetDescription& desc) {
+  desc.add<double>("HitThreshold", 0.);
+  desc.add<double>("SeedThreshold", 0.);
+  desc.add<double>("ClusterThreshold", 0.);
+}
+
+//----------------------------------------------------------------------------
+//!  Prepare the Clusterizer to work on a particular DetUnit.  Re-init the
+//!  size of the panel/plaquette (so update nrows and ncols), 
+//----------------------------------------------------------------------------
+bool MTDThresholdClusterizer::setup(const MTDGeometry* geom, const MTDTopology* topo, const DetId& id) 
+{
+  theCurrentId=id;
+  //using geopraphicalId here
+  const auto& thedet = geom->idToDet(id);
+  if( thedet == nullptr ) {
+    throw cms::Exception("MTDThresholdClusterizer") << "GeographicalID: " << std::hex
+						    << id.rawId()
+						    << " is invalid!" << std::dec
+						    << std::endl;
+    return false;
+  }
+  const ProxyMTDTopology& topoproxy = static_cast<const ProxyMTDTopology&>(thedet->topology());
+  const RectangularMTDTopology& topol = static_cast<const RectangularMTDTopology&>(topoproxy.specificTopology());    
+  
+  // Get the new sizes.
+  unsigned int nrows = topol.nrows();      // rows in x
+  unsigned int ncols = topol.ncolumns();   // cols in y
```

Comment:
related to https://github.com/cms-sw/cmssw/pull/25063/files#r243597666
perhaps it's the RectangularMTDTopology that's returning incorrect values

Reply:
No, I checked them specifically and those values are correct (nrows/ncolumns).

Occasionally for the ETL the row/column lookup returns row==nrows.
The occupancies otherwise look normal. We would prefer to debug this in a separate effort.

The rows/cols start from zero as they should.